python/lsst/daf/butler/core/datasets/ref.py

```diff
@@ -172,24 +347,36 @@ def __init__(
         id: Optional[DatasetId] = None,
         run: Optional[str] = None,
         conform: bool = True,
+        id_generation_mode: DatasetIdGenEnum = DatasetIdGenEnum.UNIQUE,
     ):
-        self.id = id
         self.datasetType = datasetType
         if conform:
             self.dataId = DataCoordinate.standardize(dataId, graph=datasetType.dimensions)
         else:
             self.dataId = dataId
-        if self.id is not None:
+        if id is not None:
             if run is None:
                 raise ValueError(
                     f"Cannot provide id without run for dataset with id={id}, "
                     f"type={datasetType}, and dataId={dataId}."
                 )
             self.run = run
+            self.id = id
         else:
             if run is not None:
-                raise ValueError("'run' cannot be provided unless 'id' is.")
-            self.run = None
+                self.run = run
+                self.id = DatasetIdFactory().makeDatasetId(
+                    self.run, self.datasetType, self.dataId, id_generation_mode
+                )
+            else:
+                self.id = None
+                self.run = None
+                warnings.warn(
+                    "Support for creating unresolved refs will soon be removed. Please contact the "
+                    "middleware team for advice on modifying your code to use resolved refs.",
+                    category=UnresolvedRefWarning,
+                    stacklevel=2,  # _find_outside_stacklevel(),
```

Comment:
Remove comment?

Reply:
Actually, I'm supposed to remove the "2". I used the 2 for testing daf_butler so I could see whether things were coming from but failed to put it back to use the right code to find the outside level -- that explains why my ci_hsc_gen3 testing is full of reports about using `ref.py` and I was surprised it wasn't telling me the real calling location outside of butler.